### DISCUSSION TITLE
[v.0.5.0] Information about the next version - juggled Onions

### DISCUSSION THREAD

**Discussion by radio24** [radio24/TorBox#119] [Archive]
Behind the scene, we (mainly nyxor, zotil and myself) have been working on different parts to publish a new main version (v.0.5.0) of TorBox (hopefully) at the end of the year. In this post, I will share what components we are working on, what has to be done to meet the deadline, and how you, as a contributor, can help us. 

**Juggled Onions**
The goal of TorBox is not only to give people an anonymizing router on a Raspberry Pi, but we also have the vision to provide a tool that helps bypass barriers such as censorship. However, censorship is not only affecting content receivers but also content deliverers. Our idea is to give out an easy to use micro-server, which allows publishing even if the publisher is behind firewalls, network address translators or lives in a censoring country. Simultaneously, this should not jeopardize the publisher's security and anonymity if correctly used. For that reason, Tor is providing Onion Services, and we started to integrate the Onion Service support into the next TorBox release.

The base for this integration founded nyxor with creating the tool [Onionservices](https://github.com/nyxnor/onionservice). So far, we have implemented the possibility to share a folder on an Onion domain (see below), but Zotil has been providing us with a TorBox File Sharing (TFS) tool which allows us to download and/or upload files as well as a TorBox Chat Secure (TCS). These two tools will also be implemented into TorBox v.0.5.0.

<img width="647" alt="OS-submenu" src="https://user-images.githubusercontent.com/46644350/142936452-938e1000-0a5d-40b1-9e1e-f87e28149eea.png">

With the above sub-menu, the user can easily create an Onion Service, and (not) publicly share a folder (with a static website):

<img width="794" alt="splash_screen_shared_folders" src="https://user-images.githubusercontent.com/46644350/142937215-6e276d13-23f6-4aa2-8c0f-cbd8e1a34ca4.png">

**Other new/improved features**
- The coming TorBox version will be based on Debian version 11 (bullseye). We will start to test it out in the following days.
- The "first use" routine will be more user-friendly and provide more options ([see here](https://github.com/radio24/TorBox/discussions/111)).
- Besides the Bridge support, we will give an option to overcome tightly configured firewalls.
- Some improvements in handling networks (especially the handling with cable connections) will be implemented based on the feedbacks of our users, and the TWM has undergone some nice improvements.
- The displayed tor logs will be decluttered, and TorBox Automatic Counter Actions (TACA) will have some new tricks up its sleeve.
- ...maybe some more

**How can you help?**
With testing and feedbacks. Tell us what is working good and what is not working. You can also contribute with code - if you see errors (even spelling errors), change them directly in the code and hand in a pull request. Also tell us, which network services you would like to see running on an Onion domain.

**How can I install TorBox v.0.5.0 ALPHA?**
Our installation script in the v.0.5.0 branch is now updated to support the installation of TorBox v.0.5.0 ALPHA?
1. Download the [latest version of the Raspberry Pi OS Lite](https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2021-11-08/2021-10-30-raspios-bullseye-armhf-lite.zip) (about 463 MB)
2. Transfer the downloaded Raspberry Pi OS Lite image on an SD Card, for example, with [Balena Etcher](https://www.balena.io/etcher/). TorBox needs at least a 8 GB SD Card.
3. Log into your newly set up system (pi/rapberry) and, if needed, configure it with “sudo raspi-config”. You need a stable Internet connection for the installation.
4. Download and run the script (to see more options, use `-h`). USE IT ONLY ON A FRESH INSTALLED SYSTEM - IT WILL CHANGE YOUR SYSTEM AND DELETE CONFIGURATIONS, FILES ETC.:
```
cd
wget https://raw.githubusercontent.com/radio24/TorBox/v.0.5.0/install/run_install.sh
chmod a+x run_install.sh  
./run_install.sh --select-branch v.0.5.0
```
5. Restart the system; after 2-3 minutes, when the green LED stops to flicker, connect your client to the new WiFi “TorBox042” (password: `CHANGE-IT`).
6. Login to the TorBox by using a SSH client (`192.168.42.1` on a WiFi client or `192.168.43.1 `on a cable client) or a web browser (`https://192.168.42.1:9000` on a WiFi client or `https://192.168.43.1:9000` on a cable client; for a connection via cable, see here; username: `torbox` / password: `CHANGE-IT`). 

**What are we doing next?**
- [x] Handle the open pull requests.
- [x] Update the installation script for the new Raspberry Pi OS (based on Debian version 11 (bullseye)).
- [x] Implementation of the TFS.
~~Implementation of the TCS.~~ --> moved to 0.5.1
- [x] Change the update-submenu so that it fits the other submenus.
- [ ] ...and more

**Comment by nyxnor** [comment]
don't forget `auth client` option, which adds clients to the `ClientOnionAuthDir /var/lib/tor/onion_auth` used for ssh over tor or any other authentication via the daemon, not the browser.

**Reply by radio24** [reply]
I'm not sure if I have enough time to integrate it to TorBox v.0.5.0. If not, it will be integrated for v.0.5.1.

**Comment by nyxnor** [comment]
getopt addition is great.

If possible to add to `run_install.sh` the option `--select-github-user` or something like it for anyone to build from their fork.

Also, instead of:
https://github.com/radio24/TorBox/blob/b4add58716c68f96e1ba5dddf5ce65c292d162ce/install/run_install.sh#L163
https://github.com/radio24/TorBox/blob/b4add58716c68f96e1ba5dddf5ce65c292d162ce/install/run_install.sh#L175
```sh
SELECT_BRANCH="--select-branch"
[ ! -z $2 ] && TORBOXMENU_BRANCHNAME="$2"
[ -z $TORBOXMENU_BRANCHNAME ] && TORBOXMENU_BRANCHNAME="master"
```
use
```sh
SELECT_BRANCH="--select-branch"
TORBOXMENU_BRANCHNAME="${2:-master}"
```

`${VAR:-STRING} ## If VAR is empty or unset, use STRING as its value.`

Read the [shell bible](https://github.com/dylanaraps/pure-sh-bible#default-value)

**Comment by radio24** [comment]
Thanks, I will add a `--select-fork `option right away.

Regarding your second suggestion: if `--select-branch` is not set, `$2` cannot be evaluated.
For example, if you start the program with `./run-install.sh` it will straight away go to `-- ) shift; break ;;` This means that the` [ -z $TORBOXMENU_BRANCHNAME ] && TORBOXMENU_BRANCHNAME="master"` has to be done **after** the while loop.

**Reply by nyxnor** [reply]
yes. good.

**Reply by radio24** [reply]
Commit: 6106aa697bb2f11c9d7f540a728761e32838f459

Done! Do you want to test it?

**Reply by nyxnor** [reply]
tested ACK.

Played with getopt, no option, one option, multiple options, working great.

I would propose instead of
```sh
# NEW v.0.5.0: We have to do that after catching the command line option
# TorBox Repository
[ -z $TORBOXMENU_FORKNAME ] && TORBOXMENU_FORKNAME="radio24"
[ -z $TORBOXMENU_BRANCHNAME ] && TORBOXMENU_BRANCHNAME="master"
TORBOXURL="https://github.com/$TORBOXMENU_FORKNAME/TorBox/archive/refs/heads/$TORBOXMENU_BRANCHNAME.zip"
```
use
```sh
TORBOXURL="https://github.com/${TORBOXMENU_FORKNAME:=radio24}/TorBox/archive/refs/heads/${TORBOXMENU_BRANCHNAME:=master}.zip"
```
Now I am using `:=` to assign the value to the variable, else it would be a one time assignment and not permanent. The `:-` would work because those variables do not appear after that line on the rest of the script.


What the `SELECT_BRANCH` and `SELECT_FORK` is used for? I presume is for the future.

I also suggest using `scurl` instead of `wget` according to https://www.whonix.org/wiki/Secure_Downloads when cloning the $TORBOXURL.
It would become
```sh
curl --tlsv1.3 --proto =https --location --remote-name ${TORBOXURL}
```

I was playing with getopt, added short options, put option arguments between `<arg>`.
```sh
# NEW v.0.5.0: New options: [-h|--help] [--select-branch branch_name]
# Catching command line options
OPTIONS=$(getopt -o h,t,f:,b:,s --long help,select-tor,select-fork:,select-branch:,step_by_step -n 'run-install' -- "$@")
if [ $? != 0 ] ; then echo "Syntax error!"; echo ""; OPTIONS="-h" ; fi
eval set -- "$OPTIONS"

SELECT_TOR=
SELECT_BRANCH=
TORBOXMENU_BRANCHNAME=
STEP_BY_STEP=
while true; do
  case "$1" in
    -h | --help )
      echo "Copyright (C) 2021 Patrick Truffer, nyxnor (Contributor)"
      echo "Syntax: run_install.sh [-h|--help] [--select-tor] [--select-branch <branch_name>] [--step_by_step]"
      echo "Options:"
      echo " -h, --help           Shows this help screen ;-)"
      echo " -t, --select-tor     Select a specific tor version (default: newest stable version)"
      echo " -f, --select-fork <fork_owner_name>"
      echo "                      Select a specific fork from a GitHub username"
      echo " -b, --select-branch <branch_name>"
      echo "                      Select a specific TorBox branch (default: master)"
      echo " -s, --step_by_step   Executes the installation step by step"
      echo ""
      echo "For more information visit https://www.torbox.ch/ or https://github.com/radio24/TorBox"
      exit 0
    ;;
    -t | --select-tor ) SELECT_TOR="--select-tor"; shift ;;
    -f | --select-fork ) TORBOXMENU_FORKNAME="${2}"; shift 2;;
    -b | --select-branch ) TORBOXMENU_BRANCHNAME="${2}"; shift 2;;
    -s | --step_by_step ) STEP_BY_STEP="--step_by_step"; shift ;;
    -- ) shift; break ;;
    * ) break ;;
  esac
done

# NEW v.0.5.0: We have to do that after catching the command line option
# TorBox Repository
TORBOXURL="https://github.com/${TORBOXMENU_FORKNAME:=radio24}/TorBox/archive/refs/heads/${TORBOXMENU_BRANCHNAME:=master}.zip"

echo "
TORBOXMENU_FORKNAME=${TORBOXMENU_FORKNAME}
TORBOXMENU_BRANCHNAME=${TORBOXMENU_BRANCHNAME}
TORBOXURL=${TORBOXURL}
"
```

**Reply by radio24** [reply]
Thanks for testing and your suggestions! Next, I'm testing the Debian installation script with Debian 11 Bullseye and Debian 12 Bookworm.

Regarding the oneliner code, I know it is shorter, and it looks geekier. However, I like to have the code readable. I'm sure, next time when I look at the code, I will say, what the heck did I here. :-D

>What the SELECT_BRANCH and SELECT_FORK is used for? I presume is for the future.

Yes, at the moment, because I also have to update the update script. If I don't need it, I will cut them out. 

Regarding `scurl`: I put it on the to-do list for v.0.5.1. I will look into it, but if I understand it right from the [Whonix website](https://www.whonix.org/wiki/Secure_Downloads), this is not a program package. It seems to be Whonix specific. Outside Whonix it seems to be replaced by `curl --tlsv1.3 --proto =https --location  --remote-name`. Unfortunately, I had some problems with curl in the past. This is why I switched to `wget` in the first place.

**Reply by nyxnor** [reply]
> Regarding the oneliner code, I know it is shorter, and it looks geekier. However, I like to have the code readable. I'm sure, next time when I look at the code, I will say, what the heck did I here. :-D

It will be readable as soon as you start using it constantly :)

Which problems with curl? Github has tlvsv1.3 but keybase.io only has 1.2, which then would fail. So `--tlsv1.2` accepts 1.2 and higher.

**Reply by radio24** [reply]
> Which problems with curl? Github has tlvsv1.3 but keybase.io only has 1.2, which then would fail. So --tlsv1.2 accepts 1.2 and higher.

I'm not sure anymore. It was at the beginning of the project on GitHub. If I remember it right, sometimes downloading from GitHub broke. I didn't have any problems with wget so far. As I said, I will take a look into it, but I cannot lose too much time with it. I'm not working in IT and starting with next week my job offers me again some other nice challenges. ;-) So, my timing for the release date at the end of the year is very tight.

**Reply by radio24** [reply]
I looked into the `curl --tlsv1.3 --proto =https --location --remote-name ${TORBOXURL}` suggestion, which is [based on a Whonix page](https://www.whonix.org/wiki/Secure_Downloads). 

Whonix claims that wget is buggy and references a post on the wget mailing list archive from July 2012. This is not the first time Whonix has pointed out an old bug, assuming that the affected software has not been developed further all these years. If I chose the used software based on past bug reports, I would probably keep changing indefinitely because bugs are normal but are usually fixed over time in active projects. In TorBox, wget runs without problems and is not used anywhere where it would jeopardize the secure use of the TorBox. At the moment, I see no reason to change the working configuration. As I wrote before, I'm short on available time.

**Comment by radio24** [comment]
We are working on the implementation of TFS.

**Reply by nyxnor** [reply]
Remember the logo I previously uploaded? It is being displayed as view only, not downloadable 
* http://l4oax3h4dgg57rz4uayfs6p7v7qvndu2ljlrdpavnemft3ygq6jupfid.onion/download/1/
* http://l4oax3h4dgg57rz4uayfs6p7v7qvndu2ljlrdpavnemft3ygq6jupfid.onion/files/home-logo.png

**Reply by nyxnor** [reply]
the mix by govnor is also doing that, and this is the default torbrowser configuration.

**Reply by nyxnor** [reply]
Is folder blocked by default? I can only put files, so this means folders have to files in  tar or zip format.

**Reply by radio24** [reply]
> Remember the logo I previously uploaded? It is being displayed as view only, not downloadable
> 
> * http://l4oax3h4dgg57rz4uayfs6p7v7qvndu2ljlrdpavnemft3ygq6jupfid.onion/download/1/
> * http://l4oax3h4dgg57rz4uayfs6p7v7qvndu2ljlrdpavnemft3ygq6jupfid.onion/files/home-logo.png

????

This is how it looks like with the Tor Browser standard installation and standard setting:

<img width="1161" alt="Bildschirmfoto 2021-12-09 um 10 44 32" src="https://user-images.githubusercontent.com/46644350/145372809-61ac6650-6893-40f1-9a04-b684e6d5dddd.png">

**Reply by nyxnor** [reply]
Yes, but clicking on media (audio, video, images) does not download them, it opens instead of downloading.

**Reply by nyxnor** [reply]
I mean, sometimes it is good to watch and see but if it is supposed to download, it is not being downloaded.
What would be better is giving to buttons, shows inline (message.txt.asc plaintext so I can read instead of downloading and media so I can see) and a second button to download.

**Reply by radio24** [reply]
> Is folder blocked by default? I can only put files, so this means folders have to files in tar or zip format.

At the moment, folders are not supported. The idea is to collect feedbacks, what people need and/or want:
- support of folders (upload/download/creation ...)?
- changes in the layout?
- upload status (but with javascript)?
- multiple instances on multiple Onion domains (currently only one instance is possible)?
- multiline description below the TFS logo?
- ....

After collecting the feedbacks and the priorities, Zotil can work on it and we will implement it in v.0.5.1.

**Reply by radio24** [reply]
> I mean, sometimes it is good to watch and see but if it is supposed to download, it is not being downloaded. What would be better is giving to buttons, shows inline (message.txt.asc plaintext so I can read instead of downloading and media so I can see) and a second button to download.

Ah, I see.... yes, you are right. Maybe we can modify it a little bit, so that clicking on it means opening and selecting means flagging for downloading. Something like this one:

<img width="883" alt="example_001" src="https://user-images.githubusercontent.com/46644350/145377526-4a7aab3e-67de-4b31-868f-1a6358f2ebbc.png">

However, I guess we need that again Java Script. What is your opinion on using JavaScript?

**Reply by nyxnor** [reply]
For each individual feature, javascript is not needed, it is possible to read plaintext or download the text file. But if integrating the buttons like I said then yes, js is needed. My opinion is that I would prefer a third option, `read-only` mode page, for seeing the video or message without downloading it. Yet another button but much less compromise than running js on the browser and node.js on your backend.

Also, if someone needs to download something from you but js is needed and their security model requires the most hardened tor browser, they will not download from your page because it will break.

**Reply by zotil** [reply]
I downloaded the DHC154 music file without problems, with a speed of 220 KB/s average. I also did upload a smaller music file for testing called: ['08.- bordemar - rin del chille.mp3'](http://l4oax3h4dgg57rz4uayfs6p7v7qvndu2ljlrdpavnemft3ygq6jupfid.onion/files/08.-_bordemar_-_rin_del_chille.mp3), uploaded fast without any issue. So seems it's working fine for uploading and downloading files.

About @nyxnor comments, I think is referencing to add the header Content-disposition as an attachment, in that case the file will not load in the browser and will download only. However, IMHO I think inside Tor network, adding more features to a web application could be turned into a problem for UX where most important thing is the privacy. So having the option to see a media file in the browser, or a text file won't harm anyone and if the user want to download can right-click and "Save link as..." (AFAIK deep web users are used to do it).

Sharing subdirectories is supported, you can have multiple subdirectories inside your TFS file path, but you can download one file at time. For downloading all files we would need to pack files and send this pack to the user, otherwise we can do other tricks without javascript to reach it, but for now I believe the main goal was reached, we can upload and download files from a path that we can choose from our TorBox (or even external devices).

Creating directories for uploading is not supported, because we also thought in the UX and privacy. Creating directories mean to have an authentication layer and identifying who is uploading. Without it, your TFS could become a mess, specially if you're running public.

We did our best to have a product without javascript and user friendly and we are open to add more features as people need. Of course there are more things that can be better and we will make it better with the time.

**Reply by radio24** [reply]
> Sharing subdirectories is supported, you can have multiple subdirectories inside your TFS file path, but you can download one file at time. 

On the TorBox side the shared directory looks like this:

<img width="804" alt="Bildschirmfoto 2021-12-09 um 17 02 19" src="https://user-images.githubusercontent.com/46644350/145432003-9332fa08-7218-495f-af1c-34bcd7f60bd0.png">

Should I see the subdirectory "test" under "Download" in TFS?

> Creating directories for uploading is not supported, because we also thought in the UX and privacy. Creating directories mean to have an authentication layer and identifying who is uploading. Without it, your TFS could become a mess, specially if you're running public.

This makes sense. Years ago, when I was running a public FTP, I always had problems: people created very looong sub-directories with special characters to try to hack the system or hide their uploaded content.

> We did our best to have a product without javascript and user friendly and we are open to add more features as people need. Of course there are more things that can be better and we will make it better with the time.

I fully agree with that. Also, feedback will help us know what is essential for the users and concentrate on these points.

**Reply by zotil** [reply]
> On the TorBox side the shared directory looks like this:
> 
> <img alt="Bildschirmfoto 2021-12-09 um 17 02 19" width="804" src="https://user-images.githubusercontent.com/46644350/145432003-9332fa08-7218-495f-af1c-34bcd7f60bd0.png">
> 
> Should I see the subdirectory "test" under "Download" in TFS?

Yes, but you need to restart the instance of TFS to re-scan the directory.

**Reply by nyxnor** [reply]
would nginx reload work?

**Reply by zotil** [reply]
> would nginx reload work?

For re-scan directory? No, TFS does the scan, so is needed to restart TFS, nginx can keep running. Maybe we can add a --scan option to the tfs command in the next versions.

**Reply by zotil** [reply]
--scan added in the #131
